### PR TITLE
Add profile menu to navbar

### PR DESCRIPTION
**Pull request by WhiteNik16**

### CODE REVIEW

File: src/common/ProfileMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  const { palette, spacing, shadows } = useTheme()
+  const { userDid } = useMetamaskZkpSnapContext()
+  const { logout } = useAuth()
+
+  let timeout: NodeJS.Timeout
+
+  const copyToClipboard = async () => {
+    try {
+      clearTimeout(timeout)
+      await navigator.clipboard.writeText(`${userDid}`)
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:41]: let's use https://www.npmjs.com/package/copy-to-clipboard to cover edge cases

File: src/common/ProfileMenu.tsx [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+      setIsCopied(true)
+      timeout = setTimeout(() => {
+        setIsCopied(false)
+      }, 3000)
+    } catch (e) {
+      bus.emit(BusEvents['error'], 'Not copied, pleas try again')
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:47]: `BusEvents.error` and typo in message

File: src/common/ProfileMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    }
+  }
+
+  const redirectToSupport = () => {
+    window.open(config.SUPPORT_LINK, '_blank', 'noopener noreferrer')
+  }
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:53]: let's use a link instead

File: src/common/ProfileMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+  const logOut = async () => {
+    await logout()
+    redirect(RoutePaths.SignIn)
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:57]: isn't redirected by auth guard?

File: src/common/ProfileMenu.tsx [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  }
+
+  const userDidLabel = useMemo(() => {
+    if (userDid.length > MAX_LENGTH_DID) {
+      return userDid.substring(0, 8) + '...' + userDid.substring(userDid.length - 4)
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:62]:
```suggestion
      return userDid.slice(0, 8) + '...' + userDid.slice(-4)
```

File: src/common/ProfileMenu.tsx [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+  const redirectToSupport = () => {
+    window.open(config.SUPPORT_LINK, '_blank', 'noopener noreferrer')
+  }
+
+  const logOut = async () => {
+    await logout()
+    redirect(RoutePaths.SignIn)
+  }
+
+  const userDidLabel = useMemo(() => {
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:60]: not label, but formateed DID

File: src/common/ProfileMenu.tsx [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    if (userDid.length > MAX_LENGTH_DID) {
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:61]: we can use a ternar here `userDid.length > MAX_LENGTH_DID ? slice... : userDid`

File: src/common/ProfileMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+import { bus } from '@/helpers'
+import { useAuth, useMetamaskZkpSnapContext } from '@/hooks'
+import { UiIcon } from '@/ui'
+
+interface ProfileMenuProps {
+  anchorEl?: null | HTMLElement | undefined
+  handleClose: () => void
+}
+
+const MAX_LENGTH_DID = 12
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:26]: what does it mean? DID cannot be lnger than 12 characters?

File: src/common/ProfileMenu.tsx [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+import { redirect } from 'react-router-dom'
+
+import { UserAvatar } from '@/common'
+import { BusEvents, RoutePaths } from '@/enums'
+import { bus } from '@/helpers'
+import { useAuth, useMetamaskZkpSnapContext } from '@/hooks'
+import { UiIcon } from '@/ui'
+
+interface ProfileMenuProps {
+  anchorEl?: null | HTMLElement | undefined
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:22]: ? already states that it can be undefined

File: src/common/ProfileMenu.tsx [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    if (userDid.length > MAX_LENGTH_DID) {
+      return userDid.substring(0, 8) + '...' + userDid.substring(userDid.length - 4)
+    }
+    return userDid
+  }, [userDid])
+
+  return (
+    <Menu
+      anchorEl={anchorEl}
+      id='account-menu'
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:70]: account or profile?

File: src/common/ProfileMenu.tsx [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+      open={open}
+      onClose={handleClose}
+      anchorOrigin={{ horizontal: 'left', vertical: 'top' }}
+      transformOrigin={{ vertical: 'bottom', horizontal: 'center' }}
+      MenuListProps={{
+        sx: {
+          width: spacing(60.5),
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:77]: let's use just 60, probably mistake in the design

File: src/common/ProfileMenu.tsx [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+          border: 1,
+          borderColor: palette.action.hover,
+          borderRadius: spacing(2),
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:80]: just 2 should work with border radius

File: src/common/ProfileMenu.tsx [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+          boxShadow: shadows[0],
+          pt: 0,
+        },
+      }}
+    >
+      <Stack direction={'column'} p={spacing(4)} alignItems={'center'}>
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:86]: p does not need spacing wrapper

File: src/common/ProfileMenu.tsx [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+        <UserAvatar sx={{ width: spacing(12), height: spacing(12) }} />
+        <Tooltip title={'Copied'} open={isCopied} onClose={handleClose}>
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:88]: we have UiTooltip now

File: src/common/ProfileMenu.tsx [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+          <Typography
+            variant={'subtitle4'}
+            overflow={'hidden'}
+            textOverflow={'ellipsis'}
+            maxWidth={spacing(30)}
+            mt={spacing(3)}
+            sx={{ cursor: 'pointer' }}
+            onClick={copyToClipboard}
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:96]: we should have additional copy button next to the DID

File: src/common/ProfileMenu.tsx [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+        <Typography variant={'caption1'}>Help Center</Typography>
+      </MenuItem>
+      <MenuItem onClick={logOut} sx={{ py: spacing(2.5), px: spacing(4) }}>
+        <ListItemIcon>
+          <UiIcon componentName={'logOut'} size={5} sx={{ color: palette.error.main }} />
+        </ListItemIcon>
+        <Typography variant={'caption1'} color={palette.error.main}>
+          Disconnect
+        </Typography>
+      </MenuItem>
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:122]: 1. extract styles to a variable
2. we don't need spacing wrapper for paddings and margins

File: src/common/ProfileMenu.tsx [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+            mt={spacing(3)}
+            sx={{ cursor: 'pointer' }}
+            onClick={copyToClipboard}
+          >
+            {userDidLabel}
+          </Typography>
+        </Tooltip>
+      </Stack>
+      <Divider sx={{ py: spacing(2) }} />
+      <MenuItem onClick={handleClose} sx={{ py: spacing(2.5), px: spacing(4) }}>
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:103]: no handler needed here at the moment, just leave a TODO

**Review by ardier16 — CHANGES_REQUESTED** [review]

File: src/common/AppNavbar.tsx [hunk 18]
```diff
@@ -91,8 +100,10 @@ const AppNavbar = () => {
             size={5}
           />
         </UiIconButton>
-        {/* TODO: add account popup */}
-        <UiIcon name={Icons.Metamask} size={10} sx={{ px: 1 }} />
+        <ProfileMenu anchorEl={anchorEl as HTMLElement} handleClose={closeProfileMenu} />
+        <UiIconButton onClick={() => openProfileMenu(event)}>
```
[the remark this answers is not in the document]
Comment (WhiteNik16) [reply src/common/AppNavbar.tsx:104]: fixed

File: src/common/AppNavbar.tsx [hunk 19]
```diff
@@ -61,6 +63,13 @@ const AppNavbar = () => {
     uiStore.setPaletteMode(paletteMode === 'dark' ? 'light' : 'dark')
   }, [paletteMode])
 
+  const openProfileMenu = (event: Event | undefined) => {
```
[the remark this answers is not in the document]
Comment (WhiteNik16) [reply src/common/AppNavbar.tsx:66]: fixed

File: src/enums/icons.ts [hunk 20]
```diff
@@ -1,3 +1,4 @@
+import { Logout, OpenInNew, QrCode } from '@mui/icons-material'
```
[the remark this answers is not in the document]
Comment (WhiteNik16) [reply src/enums/icons.ts:1]: fixed

File: src/common/UserAvatar.tsx [hunk 21]
```diff
@@ -0,0 +1,16 @@
+import { Avatar } from '@mui/material'
+import jdenticon from 'jdenticon/standalone'
+import { useMemo } from 'react'
+
+import { useMetamaskZkpSnapContext } from '@/hooks'
+
+const UserAvatar = ({ ...rest }) => {
+  const { userDid } = useMetamaskZkpSnapContext()
```
[the remark this answers is not in the document]
Comment (WhiteNik16) [reply src/common/UserAvatar.tsx:8]: fixed

File: src/common/AppNavbar.tsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (WhiteNik16) [reply src/common/AppNavbar.tsx:66]: fixed

File: src/common/ProfileMenu.tsx [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+
+export default function ProfileMenu({ anchorEl, handleClose }: ProfileMenuProps) {
+  const [isCopied, setIsCopied] = useState(false)
+
+  const open = Boolean(anchorEl)
+  const { palette, spacing, shadows } = useTheme()
+  const { userDid } = useMetamaskZkpSnapContext()
+  const { logout } = useAuth()
+
+  let timeout: NodeJS.Timeout
```
[the remark this answers is not in the document]
Comment (WhiteNik16) [reply src/common/ProfileMenu.tsx:36]: fixed

File: src/common/ProfileMenu.tsx [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+
+  const copyToClipboard = async () => {
+    try {
+      clearTimeout(timeout)
```
[the remark this answers is not in the document]
Comment (WhiteNik16) [reply src/common/ProfileMenu.tsx:40]: fixed

File: src/common/ProfileMenu.tsx (hunk 6 above)
Comment (WhiteNik16) [reply src/common/ProfileMenu.tsx:60]: fixed

File: src/common/ProfileMenu.tsx (hunk 7 above)
Comment (WhiteNik16) [reply src/common/ProfileMenu.tsx:61]: fixed

File: src/common/ProfileMenu.tsx (hunk 8 above)
Comment (WhiteNik16) [reply src/common/ProfileMenu.tsx:26]: renamed, Displayed DID length is 12 characters or less

File: src/common/ProfileMenu.tsx (hunk 10 above)
Comment (WhiteNik16) [reply src/common/ProfileMenu.tsx:70]: fixed

File: src/common/ProfileMenu.tsx (hunk 11 above)
Comment (WhiteNik16) [reply src/common/ProfileMenu.tsx:77]: fixed

File: src/common/ProfileMenu.tsx (hunk 13 above)
Comment (WhiteNik16) [reply src/common/ProfileMenu.tsx:86]: fixed

File: src/common/ProfileMenu.tsx (hunk 16 above)
Comment (WhiteNik16) [reply src/common/ProfileMenu.tsx:122]: fixed

File: src/common/ProfileMenu.tsx (hunk 15 above)
Comment (WhiteNik16) [reply src/common/ProfileMenu.tsx:96]: fixed

File: src/common/ProfileMenu.tsx (hunk 4 above)
Comment (WhiteNik16) [reply src/common/ProfileMenu.tsx:57]: fixed

File: src/common/AppNavbar.tsx [hunk 24]
```diff
@@ -47,6 +48,7 @@ const NavbarLink = ({ children, to }: NavbarLinkProps) => {
 const AppNavbar = () => {
   const { palette } = useTheme()
   const { paletteMode } = useUiState()
+  const [anchorEl, setAnchorEl] = useState<null | EventTarget | undefined>(null)
```
[the remark this answers is not in the document]
Comment (WhiteNik16) [reply src/common/AppNavbar.tsx:51]: fixed

File: src/common/UserAvatar.tsx [hunk 25]
```diff
@@ -1,14 +1,16 @@
...
 
-import { useMetamaskZkpSnapContext } from '@/hooks'
+interface UserAvatarProps extends AvatarProps {
+  userDid: string
+  size?: number
+}
 
-const UserAvatar = ({ ...rest }) => {
-  const { userDid } = useMetamaskZkpSnapContext()
+const UserAvatar = ({ userDid, size = 48, ...rest }: UserAvatarProps) => {
```
Comment (ardier16) [inline src/common/UserAvatar.tsx:10]: `size` should use spacing similar to UiIcon to be consistent

File: src/common/UserAvatar.tsx [hunk 26]
```diff
@@ -1,14 +1,16 @@
...
   const userAvatar = useMemo(() => {
-    return jdenticon.toSvg(userDid, 48)
-  }, [userDid])
+    return jdenticon.toSvg(userDid, Number(size))
```
Comment (ardier16) [inline src/common/UserAvatar.tsx:12]: already a number, no conversion is needed

File: src/common/ProfileMenu.tsx [hunk 27]
```diff
@@ -1,118 +1,108 @@
...
 interface ProfileMenuProps {
-  anchorEl?: null | HTMLElement | undefined
+  anchorEl?: HTMLElement | null
   handleClose: () => void
 }
 
-const MAX_LENGTH_DID = 12
+const MAX_LENGTH_FORMATTED_DID = 12
+
+const MenuItemStyles = { py: 2.5, px: 4 }
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:20]: menuItemSx: SxProps and put in the component

File: src/common/ProfileMenu.tsx [hunk 28]
```diff
@@ -1,118 +1,108 @@
...
+      copy(userDid)
       setIsCopied(true)
-      timeout = setTimeout(() => {
+      setTimeout(() => {
         setIsCopied(false)
-      }, 3000)
+      }, 2000)
     } catch (e) {
-      bus.emit(BusEvents['error'], 'Not copied, pleas try again')
+      bus.emit(BusEvents.error, 'Not copied, pleas try again')
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:36]:
```suggestion
      bus.emit(BusEvents.error, 'Not copied, please try again')
```

File: src/common/ProfileMenu.tsx [hunk 29]
```diff
@@ -1,118 +1,108 @@
...
   const { userDid } = useMetamaskZkpSnapContext()
   const { logout } = useAuth()
 
-  let timeout: NodeJS.Timeout
-
   const copyToClipboard = async () => {
     try {
-      clearTimeout(timeout)
-      await navigator.clipboard.writeText(`${userDid}`)
+      copy(userDid)
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:30]: no error is thrown here, instead it returns `success` boolean

File: src/common/ProfileMenu.tsx [hunk 30]
```diff
@@ -1,118 +1,108 @@
...
-            mt={spacing(3)}
-            sx={{ cursor: 'pointer' }}
-            onClick={copyToClipboard}
           >
-            {userDidLabel}
+            {formatedDidLabel}
           </Typography>
-        </Tooltip>
+          <UiTooltip title={'Copied'} open={isCopied} onClose={handleClose}>
+            <UiIconButton onClick={copyToClipboard} sx={{ ml: 2 }}>
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:83]: use spacing on stack instead of `ml`

File: src/common/ProfileMenu.tsx [hunk 31]
```diff
@@ -1,118 +1,108 @@
...
         </ListItemIcon>
         <Typography variant={'caption1'}>QR login</Typography>
       </MenuItem>
-      <MenuItem onClick={redirectToSupport} sx={{ py: spacing(2.5), px: spacing(4) }}>
-        <ListItemIcon>
-          <UiIcon componentName={'openInNew'} size={5} sx={{ color: palette.text.secondary }} />
-        </ListItemIcon>
-        <Typography variant={'caption1'}>Help Center</Typography>
+      <MenuItem sx={MenuItemStyles}>
+        <Link to={config.SUPPORT_LINK} target='_blank' rel='noreferrer noopener'>
```
Comment (ardier16) [inline src/common/ProfileMenu.tsx:98]: you can try `<MenuItem component={Link} to=...` to not produce multiple elements

File: src/hooks/auth.ts [hunk 32]
```diff
@@ -21,25 +22,29 @@ export const useAuth = () => {
     checkSnapStatus,
     createIdentity,
   } = useMetamaskZkpSnapContext()
+  // Todo: isConnected change to jwt. We can simulate a disconnect by removing jwt
```
Comment (ardier16) [inline src/hooks/auth.ts:25]: user could be logged in without JWT, JWT is needed for org management

**Review by ardier16 — CHANGES_REQUESTED** [review]

File: src/common/UserAvatar.tsx (hunk 26 above)
Comment (WhiteNik16) [reply src/common/UserAvatar.tsx:12]: now it is converted with "spacing" which returns string, but the function accepts type number

**Review by lukachi — CHANGES_REQUESTED** [review]

File: src/common/ProfileMenu.tsx [hunk 33]
```diff
@@ -0,0 +1,104 @@
...
+            <Typography
+              variant={'subtitle4'}
+              overflow={'hidden'}
+              textOverflow={'ellipsis'}
+              maxWidth={spacing(30)}
+            >
+              {formatDid(userDid)}
+            </Typography>
+            <UiTooltip title={'Copied'} open={isCopied && !!anchorEl}>
+              <UiIconButton onClick={() => copyToClipboard(userDid)}>
```
Comment (lukachi) [inline src/common/ProfileMenu.tsx:67]: add method handler

File: src/common/UserAvatar.tsx [hunk 34]
```diff
@@ -0,0 +1,21 @@
...
+  size?: number
+}
+
+const UserAvatar = ({ userDid, size = 12, ...rest }: UserAvatarProps) => {
+  const { spacing } = useTheme()
+
+  const userAvatar = useMemo(() => {
+    return jdenticon.toSvg(userDid, parseInt(spacing(size)))
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [size, userDid])
```
Comment (ardier16) [inline src/common/UserAvatar.tsx:16]: probably spacing should here as well

File: src/hooks/copy.ts [hunk 35]
```diff
@@ -0,0 +1,20 @@
+import copy from 'copy-to-clipboard'
+import { useState } from 'react'
+
+import { sleep } from '@/helpers'
+
+export const useCopy = () => {
```
Comment (ardier16) [inline src/hooks/copy.ts:6]: `useCopyToClipboard`

File: src/hooks/copy.ts [hunk 36]
```diff
@@ -0,0 +1,20 @@
...
+  const [isCopied, setIsCopied] = useState(false)
+  //Todo: migrate to navigator Api
+  const copyToClipboard = async (value: string) => {
```
Comment (ardier16) [inline src/hooks/copy.ts:9]: IMO just `copy` is enough

add useCallback

File: src/helpers/format.ts [hunk 37]
```diff
@@ -0,0 +1,7 @@
+const MAX_LENGTH_FORMATTED_DID = 12
+
+export function formatDid(userDid: string) {
```
Comment (ardier16) [inline src/helpers/format.ts:3]: just did, not userDid

File: src/helpers/format.ts [hunk 38]
```diff
@@ -0,0 +1,7 @@
+const MAX_LENGTH_FORMATTED_DID = 12
```
Comment (ardier16) [inline src/helpers/format.ts:1]: FORMATTED_DID_MAX_LENGTH

**Review by ardier16 — CHANGES_REQUESTED** [review]

**Review by lukachi — APPROVED** [review]

**Review by ardier16 — APPROVED** [review]

**Pull request merged**